executor/runtime/docker/docker.go

```diff
@@ -1484,15 +1478,60 @@ func (r *DockerRuntime) Start(parentCtx context.Context, pod *v1.Pod) (string, *
 		return "", nil, statusMessageChan, err
 	}
 	mainContainerRoot := getMainContainerRoot(inspectOutput)
-	err = r.startUserContainers(ctx, pod, r.c.ID(), tiniConn, mainContainerRoot)
+
+	allOtherContainerIDs := r.getExtraContainerIDs()
+	if len(allOtherContainerIDs) > 0 {
+		err = r.createOtherContainers(ctx, pod, r.c.ID(), tiniConn, mainContainerRoot)
+		if err != nil {
+			eventCancel()
+			return "", nil, statusMessageChan, err
+		}
+		// This second docker events channel is for the extra containers, and only exists if we have other containers
+		// to work with. This allows the main container channed (created way earliar) to start and watch way before
+		// the sidecar containers even exist.
+		eventChanExtra, eventErrChanExtra := r.getDockerEventsChannelsForContainers(eventCtx, allOtherContainerIDs)
+		for _, c := range r.c.ExtraPlatformContainers() {
+			entry.Debugf("Adding status monitor for %s container (cid %s)", c.Name, c.Status.ContainerID)
+			go r.statusMonitor(eventCancel, c.Status.ContainerID, eventChanExtra, eventErrChanExtra, statusMessageChan)
+		}
+		entry.Infof("Starting %d platform sidecars: %v", len(r.getPlaformContainerNames()), r.getPlaformContainerNames())
+		err = r.startPlatformDefinedContainers(ctx)
+		if err != nil {
+			eventCancel()
+			return "", nil, statusMessageChan, fmt.Errorf("Failed to start a platform-sidecar container: %s", err)
+		}
+		entry.Infof("Starting %d user-defined containers: %v", len(r.getUserContainerNames()), r.getUserContainerNames())
+		for _, c := range r.c.ExtraUserContainers() {
+			entry.Debugf("Adding status monitor for %s container (cid %s)", c.Name, c.Status.ContainerID)
+			go r.statusMonitor(eventCancel, c.Status.ContainerID, eventChanExtra, eventErrChanExtra, statusMessageChan)
+		}
+	}
+
+	// Last, we actually start the containers. And by start we mean:
+	// platform sidecars first (via docker start)
+	// user extra containers second (via docker start)
+	// main container (via tini)
+	err = r.startUserDefinedContainers(ctx, tiniConn)
 	if err != nil {
 		eventCancel()
-		return "", nil, statusMessageChan, err
+		return "", nil, statusMessageChan, fmt.Errorf("Failed to start a user-defined container: %s", err)
 	}
 
 	return logDir, details, statusMessageChan, err
 }
 
+func (r *DockerRuntime) getDockerEventsChannelsForContainers(eventCtx context.Context, containerIDs []string) (<-chan events.Message, <-chan error) {
+	filters := filters.NewArgs()
+	for _, containerID := range containerIDs {
+		filters.Add("container", containerID)
+		filters.Add("type", "container")
```

Comment:
Shouldn't this be outside the loop?

Reply:
I don't think so. Seems to work correctly while testing. 
Per the docs (at least with cli):
https://docs.docker.com/engine/reference/commandline/events/#limiting-filtering-and-formatting-the-output
Not sure if it is redundant or not. The tests should tell though, let me see for sure.